### PR TITLE
Metadata composer library

### LINKED ISSUE #2: A library to aggregate metadata from various networks
This library will give users the options to configure subgraph urls where metadata from different networks can be queried.

The input will be GraphQL query.
The output will be encoded as JSON format, similar to what theGraph api would return.
The output will be aggregated and sorted

### PR DESCRIPTION
**Pull request by yuetloo**
Currently, the library merges the subgraph results from different networks.  It doesn't do sorting, which is documented in issue #8 .

This closes #2

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -23,10 +23,12 @@ Subgraph is used to track the state of metadata by indexing events logged throug
 
 Content of the metadata will not be verified as the registry is designed to be generic and can accept any metadata. It is the Dapp's responsibility to verify metadata.
 
-See subgraph entity definitions: [/subgraph/schema.graphql](./subgraph/schema.graphql).
-
+See subgraph entity definitions: [/packages/subgraph/schema.graphql](./packages/subgraph/schema.graphql).
```
Comment (daodesigner) [inline README.md:26]: generate UML from schema for the docs: 
https://github.com/sheerun/graphqlviz or
https://github.com/APIs-guru/graphql-voyager

File: packages/composer/README.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+//          id: 'rinkeby-0x123...456-59'
+//        }
+//      ]
+//    }
+//
+```
+
+## Limitations
+
+Currently, this library only merges the results without sorting or refining the paging according to the query request.
```
Comment (daodesigner) [inline packages/composer/README.md:29]: open issues to track these

File: packages/composer/README.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+## Usage
+```
+import { MetadataComposer } from '@clrfund/metadata-composer'
+
+const urls = ['https://api.thegraph.com/subgraphs/name/yuetloo/metadata-rinkeby']
+const composer = new MetadataComposer(urls)
+
+const query = '{ metadataEntries{id} }'
+const result = await composer.query(query)
```
Comment (daodesigner) [inline packages/composer/README.md:13]: return FE friendly responses, type out the response to make error handling explicit
ie. `const { data, loading, error } = await composer.query(query)`

File: packages/composer/package.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "@clrfund/composer",
+  "version": "1.0.0",
```
Comment (daodesigner) [inline packages/composer/package.json:3]: start at sub 1.0.0 version

File: packages/composer/src/MetadataComposer.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import { createClient, Client, OperationResult } from '@urql/core'
```
Comment (daodesigner) [inline packages/composer/src/MetadataComposer.ts:1]: love urql <3

File: packages/composer/src/MetadataComposer.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import fetch from 'isomorphic-unfetch'
+
+/**
+ * Merge an array of subgraph query results from different networks
+ * Export this function for testing purposes
+ * @param result array of query results from subgraphs
+ * @returns merged result
+ */
+export function merge(result: OperationResult<any, any>[]): any {
```
Comment (daodesigner) [inline packages/composer/src/MetadataComposer.ts:10]: should merge according to a time stamp? should be an order.

File: packages/composer/src/MetadataComposer.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+   * @param queryString graphql query
+   * @param args arguments to pass to the query
+   * @returns query result
+   */
+  async query(queryString: string, args: any = {}): Promise<any> {
+    const queryResult = await Promise.all(
+      this.clients.map((client) => client.query(queryString, args).toPromise())
+    )
+
+    const error = queryResult.find((res) => res.error)
```
Comment (daodesigner) [inline packages/composer/src/MetadataComposer.ts:53]: only returns the first error encountered, should ( optionally? ) return all the errors encountered if there is a debug flag or similar.

File: packages/composer/src/MetadataComposer.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    if (error) {
+      return error
+    }
+
+    return merge(queryResult)
```
Comment (daodesigner) [inline packages/composer/src/MetadataComposer.ts:58]: don't assume this cannot error out, prefer try catch blocks to explicitly handle errors

File: packages/composer/test/composer.test.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+  }
+}
+
+function createPermission(key: number) {
+  return {
+    id: `p${key}`,
+  }
+}
+
+describe('MetadataComposer', () => {
```
Comment (daodesigner) [inline packages/composer/test/composer.test.ts:36]: create issue to use GitHub Actions to Run Automated Tests on PR submission

File: packages/composer/tsconfig.json [hunk 10]
```diff
@@ -0,0 +1,7 @@
+{
+  "extends": "./tsconfig.base.json",
+  "compilerOptions": {
+    "noEmit": true
```
Comment (daodesigner) [inline packages/composer/tsconfig.json:4]: add types

**Review by daodesigner — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (yuetloo) [reply README.md:26]: added UML as requested.

File: packages/composer/README.md (hunk 2 above)
Comment (yuetloo) [reply packages/composer/README.md:29]: issue for this: #8

File: packages/composer/src/MetadataComposer.ts (hunk 6 above)
Comment (yuetloo) [reply packages/composer/src/MetadataComposer.ts:10]: the technical issue is that we might not have a timestamp to order by, depending on what fields come back from the result.

File: packages/composer/test/composer.test.ts (hunk 9 above)
Comment (yuetloo) [reply packages/composer/test/composer.test.ts:36]: created issue #10 for this.

File: packages/composer/tsconfig.json (hunk 10 above)
Comment (yuetloo) [reply packages/composer/tsconfig.json]: added types in tsconfig.json, please re-review.

File: packages/composer/README.md (hunk 3 above)
Comment (yuetloo) [reply packages/composer/README.md]: done.

File: packages/composer/src/MetadataComposer.ts (hunk 8 above)
Comment (yuetloo) [reply packages/composer/src/MetadataComposer.ts]: added the try catch block

**Review by daodesigner — APPROVED** [review]

**Pull request merged**